ament_cmake_pytest/cmake/ament_add_pytest_test.cmake

```diff
@@ -98,6 +98,46 @@ function(ament_add_pytest_test testname path)
     list(APPEND cmd "-We")
   endif()
 
+  option(AMENT_CMAKE_PYTEST_WITH_COVERAGE
+    "Generate coverage information for Python tests"
+    OFF)
+
+  # enable pytest coverage automatically if the package test_depends on pytest_cov
+  if("python3-pytest-cov" IN_LIST ${PROJECT_NAME}_TEST_DEPENDS)
+    set(AMENT_CMAKE_PYTEST_WITH_COVERAGE ON)
+  endif()
+
+  if(AMENT_CMAKE_PYTEST_WITH_COVERAGE)
+    # check if pytest-cov is available
+    ament_has_pytest_cov(has_pytest_cov pytest_cov_version
+      QUIET PYTHON_EXECUTABLE "${ARG_PYTHON_EXECUTABLE}"
+    )
+    if(NOT has_pytest_cov)
+      message(WARNING
+        "The Python module 'pytest-cov' was not found, test coverage will not be produced "
+        "(e.g. on Ubuntu/Debian install the package 'python3-pytest-cov')")
+      return()
+    else()
+      list(APPEND cmd
+        "--cov=${CMAKE_CURRENT_SOURCE_DIR}"
+        "--cov-report=html:${CMAKE_CURRENT_BINARY_DIR}/coverage.html"
+        "--cov-report=xml:${CMAKE_CURRENT_BINARY_DIR}/coverage.xml"
+        "--cov-append"
```

Comment:
I don't this option makes sense. That would append the result from multiple test runs which seems to be undesired.

Reply:
You're right. I was testing with `tracetools_test`, which (for some reason that I don't remember) [does multiple calls to `ament_add_pytest_test()`](https://gitlab.com/micro-ROS/ros_tracing/ros2_tracing/-/blob/master/tracetools_test/CMakeLists.txt#L109), so without that option, we only get the results from the last test. I guess this isn't a normal use-case.

Fixed in 634e79be0695bd61e4a747ca68b7d058363546f8.

Reply:
Actually a package might add multiple `pytest` test targets. In that case they would all try to write the same coverage result files.

Maybe it is best to place the `coverage.*` files into a subdirectory named after the test?

Reply:
Done in 2334fbeddae8d66004ab85a7168eceadd1c10518.

Reply:
I have just tried the patch and it fails during running the tests. I would guess because the directory where the coverage are supposed to be generated doesn't exist.

Please make sure to run / test your patch when making changes.

Reply:
@dirk-thomas could you provide more information? Are you on Focal?

I tested both use-cases (with many tests in the same `ament_add_pytest_test()` call & one call to `ament_add_pytest_test()` per test) a couple times, even deleting my package's `build` directoy before building, and it worked fine.

Reply:
@dirk-thomas I just tested on Focal (with `pytest-cov` `2.8.1` as opposed to `2.7.1` on Bionic) and I indeed see the failures. I didn't get that on Bionic. Will fix.

Reply:
Fixed in c41424037fa5b0dda57fb68cfea623fdf8bf1e18.

Sorry, I should've tested on Focal too.